ec/ecresource/deploymentresource/update.go

```diff
@@ -19,73 +19,150 @@ package deploymentresource
-	client := meta.(*api.API)
+func (r *Resource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var plan v2.DeploymentTF
 
-	if hasDeploymentChange(d) {
-		if err := updateDeployment(ctx, d, client); err != nil {
-			return diag.FromErr(err)
-		}
-	}
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
 
-	if err := handleTrafficFilterChange(d, client); err != nil {
-		return diag.FromErr(err)
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
-	if err := handleRemoteClusters(d, client); err != nil {
-		return diag.FromErr(err)
+	var state v2.DeploymentTF
+
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
-	return readResource(ctx, d, meta)
-}
+	updateReq, diags := plan.UpdateRequest(ctx, r.client, state)
 
-func updateDeployment(_ context.Context, d *schema.ResourceData, client *api.API) error {
-	req, err := updateResourceToModel(d, client)
-	if err != nil {
-		return err
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
 	}
 
 	res, err := deploymentapi.Update(deploymentapi.UpdateParams{
-		API:          client,
-		DeploymentID: d.Id(),
-		Request:      req,
+		API:          r.client,
+		DeploymentID: plan.Id.Value,
+		Request:      updateReq,
 		Overrides: deploymentapi.PayloadOverrides{
-			Version: d.Get("version").(string),
-			Region:  d.Get("region").(string),
+			Version: plan.Version.Value,
+			Region:  plan.Region.Value,
 		},
 	})
 	if err != nil {
-		return multierror.NewPrefixed("failed updating deployment", err)
+		resp.Diagnostics.AddError("failed updating deployment", err.Error())
+		return
 	}
 
-	if err := WaitForPlanCompletion(client, d.Id()); err != nil {
-		return multierror.NewPrefixed("failed tracking update progress", err)
+	if err := WaitForPlanCompletion(r.client, plan.Id.Value); err != nil {
+		resp.Diagnostics.AddError("failed tracking update progress", err.Error())
+		return
 	}
 
-	return parseCredentials(d, res.Resources)
+	resp.Diagnostics.Append(handleTrafficFilterChange(ctx, r.client, plan, state)...)
+
+	resp.Diagnostics.Append(v2.HandleRemoteClusters(ctx, r.client, plan.Id.Value, plan.Elasticsearch)...)
+
+	deployment, diags := r.read(ctx, plan.Id.Value, &state, plan, res.Resources)
+
+	resp.Diagnostics.Append(diags...)
+
+	if deployment == nil {
+		resp.Diagnostics.AddError("cannot read just updated resource", "")
+		resp.State.RemoveResource(ctx)
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, deployment)...)
 }
 
-// hasDeploymentChange checks if there's any change in the resource attributes
-// except in the "traffic_filter" prefixed keys. If so, it returns true.
-func hasDeploymentChange(d *schema.ResourceData) bool {
-	for attr := range d.State().Attributes {
-		if strings.HasPrefix(attr, "traffic_filter") {
-			continue
+func handleTrafficFilterChange(ctx context.Context, client *api.API, plan, state v2.DeploymentTF) diag.Diagnostics {
```

Comment:
Would be great to have some unit tests for these

Reply:
I added tests.